pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1459 @@
+				Error::<T>::Forbidden
+			);
+
+			let position = Positions::<T>::get(position_id).ok_or(Error::<T>::PositionNotFound)?;
+
+			ensure!(position.shares >= amount, Error::<T>::InsufficientShares);
+
+			let stable_asset = Self::stable_asset()?;
+
+			let asset_id = position.asset_id;
+
+			let asset_state = Self::load_asset_state(asset_id)?;
+
+			ensure!(
+				asset_state.tradable.contains(Tradability::REMOVE_LIQUIDITY),
+				Error::<T>::NotAllowed
+			);
+
+			let state_changes = hydra_dx_math::omnipool::calculate_remove_liquidity_state_changes(
+				&(&asset_state).into(),
+				amount,
+				&(&position).into(),
+				stable_asset,
+				asset_id == T::StableCoinAssetId::get(),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			let new_asset_state = asset_state
+				.delta_update(&state_changes.asset)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Update position state
+			let updated_position = position
+				.delta_update(
+					&state_changes.delta_position_reserve,
+					&state_changes.delta_position_shares,
+				)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Token balance updates
+			T::Currency::transfer(
+				asset_id,
+				&Self::protocol_account(),
+				&who,
+				*state_changes.asset.delta_reserve,
+			)?;
+
+			let delta_imbalance = Self::recalculate_imbalance(&new_asset_state, state_changes.delta_imbalance)
+				.ok_or(ArithmeticError::Overflow)?;
+			Self::update_imbalance(<HubAssetImbalance<T>>::get(), delta_imbalance)?;
+
+			Self::update_tvl(&state_changes.asset.delta_tvl)?;
+
+			Self::update_hub_asset_liquidity(&state_changes.asset.delta_hub_reserve)?;
+
+			// LP receives some hub asset
+			if state_changes.lp_hub_amount > Balance::zero() {
+				T::Currency::transfer(
+					T::HubAssetId::get(),
+					&Self::protocol_account(),
+					&who,
+					state_changes.lp_hub_amount,
+				)?;
+			}
+
+			if updated_position.shares == Balance::zero() {
+				// All liquidity removed, remove position and burn NFT instance
+
+				<Positions<T>>::remove(position_id);
+				T::NFTHandler::burn_from(&T::NFTClassId::get(), &position_id)?;
+
+				Self::deposit_event(Event::PositionDestroyed {
+					position_id,
+					owner: who.clone(),
+				});
+			} else {
+				<Positions<T>>::insert(position_id, updated_position);
+			}
+
+			Self::set_asset_state(asset_id, new_asset_state);
+
+			Self::deposit_event(Event::LiquidityRemoved {
+				who,
+				position_id,
+				asset_id,
+				shares_removed: amount,
+			});
+
+			Ok(())
+		}
+
+		/// Sacrifice LP position in favor of pool.
+		///
+		/// A position is destroyed and liquidity owned by LP becomes pool owned liquidity.
+		///
+		/// Only owner of position can perform this action.
+		///
+		/// Emits `PositionDestroyed`.
+		#[pallet::weight(<T as Config>::WeightInfo::sacrifice_position())]
+		#[transactional]
+		pub fn sacrifice_position(origin: OriginFor<T>, position_id: T::PositionInstanceId) -> DispatchResult {
```

Comment:
what is the usecase for this? i could imaging if the initialization of omnipool would be creating position this could be then used, but still i would prefer to have option to contol the position

Reply:
@poliwop  can tell the exact motivation to have ability to give up a position in favor of pool. 

the initial liquidity is already owned by protocol, the sacrifice position adds shares of position to protocol shares.

Reply:
where are protocol shares tracked?

Reply:
> where are protocol shares tracked?

in state of each asset.